### DISCUSSION TITLE
How to debug ceph rgw memory leak in production env?

### DISCUSSION THREAD

**Discussion by microyahoo** [rook/rook#13119] [Q&A]
After running the ceph cluster for a period of time, I found that the rgw memory usage of several nodes was particularly high. So my question is how to debug or figure out ceph rgw memory leak?
![WX20231028-224828@2x](https://github.com/rook/rook/assets/10202906/252b7c67-76a1-4f21-9f7d-c9352c0ca370)
![WX20231028-225129@2x](https://github.com/rook/rook/assets/10202906/d7545ca2-bebf-472f-bff0-ffa720497ea8)

**Comment by travisn** [comment]
Have you set the rgw [resource requests and limits](https://github.com/rook/rook/blob/release-1.12/deploy/examples/object.yaml#L94-L101)?

**Reply by microyahoo** [reply]
no, rgw pods have not set resource limit.

**Reply by travisn** [reply]
Any reason not to set them? The limits would help avoid such the high usage.

**Reply by microyahoo** [reply]
We have previously discovered that memory limits have an impact on performance, and our cluster memory resources are sufficient, so we have no restrictions on rgw pods.
For the above, how to figure out whether it is a memory leak or whether so much memory is really needed, and how to debug it? Thanks!

**Reply by thenamehasbeentake** [reply]
Here is [ceph memory profile](https://docs.ceph.com/en/latest/rados/troubleshooting/memory-profiling/#memory-profiling) document which only supports  mon, osd and mds daemon.

**Reply by travisn** [reply]
What version of Ceph are you running? 
You could try tuning the [rgw_max_concurrent_requests](https://docs.ceph.com/en/latest/radosgw/config-ref/#confval-rgw_max_concurrent_requests) or [rgw_thread_pool_size](https://docs.ceph.com/en/latest/radosgw/config-ref/#confval-rgw_thread_pool_size). Try setting them to the same value, 1024 or 2048, but no higher.

**Reply by microyahoo** [reply]
ceph pacific 16.2.13

**Reply by travisn** [reply]
You might consider upgrading to quincy, though I don't know if there are specific fixes for rgw memory.

**Reply by microyahoo** [reply]
Got it, thanks.
